java/ql/src/experimental/Security/CWE/CWE-601/SpringUrlRedirect.qll

```diff
@@ -0,0 +1,80 @@
+import java
+import DataFlow
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.DataFlow2
+import semmle.code.java.dataflow.TaintTracking
+import semmle.code.java.frameworks.spring.SpringController
+
+class StartsWithSanitizer extends DataFlow::BarrierGuard {
+  StartsWithSanitizer() {
+    this.(MethodAccess).getMethod().hasName("startsWith") and
+    this.(MethodAccess).getMethod().getDeclaringType() instanceof TypeString and
+    this.(MethodAccess).getMethod().getNumberOfParameters() = 1
+  }
+
+  override predicate checks(Expr e, boolean branch) {
+    e = this.(MethodAccess).getQualifier() and branch = true
+  }
+}
+
+/**
+ * A concatenate expression using the string `redirect:` on the left.
+ *
+ * E.g: `"redirect:" + redirectUrl`
+ */
+class RedirectBuilderExpr extends AddExpr {
+  RedirectBuilderExpr() {
+    this.getLeftOperand().(CompileTimeConstantExpr).getStringValue() = "redirect:"
+  }
+}
+
+/** A URL redirection sink from spring controller method. */
+class SpringUrlRedirectSink extends DataFlow::Node {
+  SpringUrlRedirectSink() {
+    exists(RedirectBuilderExpr rbe | rbe.getRightOperand() = this.asExpr())
```

Comment:
Shouldnt we make sure the "redirect:" prefixed string flows into the return expression of a controller method?

Reply:
This question @smowton mentioned.

Reply:
I think @smowton refered to the `RedirectView` and `setUrl` cases where these objects are used only for constructing Views so they will probably be used. But just concatenating "redirect:" is different since its very generic, you may even find it on non-Spring applications

Reply:
Ah damn you're right, I did suggest `redirect: + <sink>` as itself forming a sink, to avoid the usual pitfalls of daisy-chaining two global dataflow configs together. How about using that as a sink + ensuring we're looking at a Spring application by checking a relevant Spring type exists in the database? In such a context would you consider `"redirect:" + <sink>` sufficiently suspicious?

Reply:
@smowton How about checking that the enclosing callable is reached from a Spring controller method, something along the lines of

```ql
exists(Method controller_method |
	controller_method.getAnAnnotation().getType().getName() = ["RequestMapping", "GetMapping", "PostMapping", ...] and
    controller_method.getACallee*() = sink.getEnclosingCallable()
)
```

Reply:
```ql
exists(Method controller_method |
        controller_method.getAnAnnotation().getType().getName() = ["RequestMapping", "GetMapping", "PostMapping"] and
        any(controller_method).polyCalls*(sink.getEnclosingCallable())
)
```

@pwntester This should be what you want.

Reply:
That `any` isn't doing anything.

How about

```
any(Method controller_method | controller_method.getAnAnnotation().getType().getName() = ["RequestMapping", "GetMapping", "PostMapping"]).polyCalls*(sink.getEnclosingCallable())
```

Reply:
Cool, just learnt about `polyCalls` is it semantically similar to what I proposed?

Reply:
Yes but it traverses virtual dispatch too (e.g. `Object.toString` perhaps actually calling `String.toString`)